src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogBuildListener.java

```diff
@@ -114,10 +114,16 @@ public void onInitialize(Run run) {
                 return;
             }
 
-            final TraceSpan buildSpan = new TraceSpan("jenkins.build", TimeUnit.MILLISECONDS.toNanos(buildData.getStartTime(0L)));
-            BuildSpanManager.get().put(buildData.getBuildTag(""), buildSpan);
+            TraceSpan.TraceSpanContext buildSpanContext = new TraceSpan.TraceSpanContext();
+            BuildSpanManager.get().put(buildData.getBuildTag(""), buildSpanContext);
 
-            final BuildSpanAction buildSpanAction = new BuildSpanAction(buildSpan.context());
+            TraceSpan.TraceSpanContext upstreamBuildSpanContext = null;
+            String upstreamBuildTag = buildData.getUpstreamBuildTag("");
+            if (upstreamBuildTag != null) {
+                upstreamBuildSpanContext = BuildSpanManager.get().get(upstreamBuildTag);
```

Comment:
So basically here, we're assuming the `upstream pipeline` has been initialised before the `downstream` pipeline, and we try to find the Upstream pipeline in the Manager, is that correct?
Maybe we should add some brief comments explaining this. 

Additionally, if there's `upstreamBuildTag`, but the returned `upstreamBuildSpanContext` is `null`, it'd mean that the upstream context has been removed from the queue, and I think we should log something in this case. This might happen that your upstream pipeline takes a lot of time, and you have a lot of short pipelines in the middle that can exhaust the in-memory cache.

The solution for that would be to increase the cache size using the environment variable, if I'm not wrong. That's the reason I think we should log some hint for customers.

Reply:
> upstream pipeline has been initialised before the downstream pipeline

Yes, that is the definition of "upstream" I think :)
Added a comment and a log with the hint.